multinode_integration_tests/src/main.rs

```diff
@@ -342,6 +349,104 @@ impl MockNode {
     }
 }
 
+#[derive(Debug, Eq, PartialEq, Clone)]
+struct ProbeTarget {
+    ip_address: IpAddr,
+    port_specs: Vec<PortSpec>,
+}
+
+impl FromStr for ProbeTarget {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut main_pieces = s.split(':').collect_vec();
+        if main_pieces.len() != 2 {
+            return Err("Syntax: <IP address>:<U|T><port>/<U|T><port>/...".to_string());
+        }
+        let ip_address_string = main_pieces.remove(0);
+        let port_specs_string = main_pieces.remove(0);
+        let ip_address = match IpAddr::from_str(ip_address_string) {
+            Ok(ip_addr) => ip_addr,
+            Err(_) => return Err("Syntax: <IP address>:<U|T><port>/<U|T><port>/...".to_string()),
+        };
+        let port_spec_strings = port_specs_string.split('/').collect_vec();
+        let port_spec_results = port_spec_strings
+            .into_iter()
+            .map(PortSpec::from_str)
+            .collect_vec();
+        let first_error_opt = port_spec_results
+            .iter()
+            .find(|result| result.is_err())
+            .map(|err| err.clone().err().unwrap());
+        if first_error_opt.is_some() {
+            Err("Syntax: <IP address>:<U|T><port>/<U|T><port>/...".to_string())
+        } else {
+            let port_specs = port_spec_results
+                .into_iter()
+                .map(|result| result.unwrap())
+                .collect_vec();
+            Ok(ProbeTarget {
+                ip_address,
+                port_specs,
+            })
+        }
+    }
+}
+
+impl Display for ProbeTarget {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let port_spec_string = self
+            .port_specs
+            .iter()
+            .map(|port_spec| format!("{}{}", port_spec.protocol, port_spec.port))
+            .join("/");
+        write!(f, "{}:{}", self.ip_address, port_spec_string)
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+struct PortSpec {
+    protocol: NetworkProtocol,
+    port: u16,
+}
+
+impl FromStr for PortSpec {
+    type Err = String;
```

Comment:
Again, `&'static str` is an attractive option here. Shaking off the `.to_string()`s.

Reply:
Okay...